### PR TITLE
Fix platform guard analysis with call-site support

### LINKED ISSUE #7661: CA1416 with guard attributes when non-taken path has UnsupportedOSPlatform attribute
### Analyzer

**Diagnostic ID**: warning CA1416: This call site is reachable on: 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is unsupported on: 'macOS/OSX' all versions.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

```
$ ./.dotnet/dotnet --version
10.0.100-preview.3.25167.3
```

### Describe the bug

Incorrect CA1416 when trying to access API in code block protected using guard attributes, and the path not taken has code with UnsupportedOSPlatform attributes.

```cs
using System;
using System.Runtime.Versioning;

[assembly: SupportedOSPlatform ("macos12.0")]

partial class TestType {
    [SupportedOSPlatform ("macos12.0")]
    public virtual object DictionaryRepresentation {
        get {
            if (IsAtLeastXcode12) {
                return _DictionaryRepresentation14;
            } else {
                return _DictionaryRepresentation13; // warning CA1416: This call site is reachable on: 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is unsupported on: 'macOS/OSX' all versions.
            }
        }
    }

    [SupportedOSPlatformGuard ("macos")]
    public ulong? _DictionaryRepresentation14 { get; private set; }

    [Mock.UnsupportedOSPlatform ("macos")]
    public ulong? _DictionaryRepresentation13 { get; private set; }

    [SupportedOSPlatformGuard ("macos12.0")]
    internal static bool IsAtLeastXcode12 {
        get {
            return true;
        }
    }
}
```

Or as a unit test: https://github.com/rolfbjarne/roslyn-analyzers/commit/48ebc77ea21e147c8e0d6c69bbbb57f9f490274c

Reports:

### Steps To Reproduce

Run this unit test: https://github.com/rolfbjarne/roslyn-analyzers/commit/48ebc77ea21e147c8e0d6c69bbbb57f9f490274c

### Expected behavior

Unit test should pass.

### Actual behavior

Unit test fails:

```
System.InvalidOperationException : Context: Diagnostics of test state
Mismatch between number of diagnostics returned, expected "0" actual "1"

Diagnostics:
// /0/Test0.cs(16,24): warning CA1416: This call site is reachable on: 'ios' 14.0 and later, 'maccatalyst' 14.0 and later, 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is only supported on: 'ios' 12.2 and later, 'maccatalyst' 12.2 and later.
VerifyCS.Diagnostic(PlatformCompatibilityAnalyzer.CA1416).WithSpan(16, 24, 16, 51).WithArguments("TestType._DictionaryRepresentation13", "'ios' 12.2 and later, 'maccatalyst' 12.2 and later", "'ios' 14.0 and later, 'maccatalyst' 14.0 and later, 'macOS/OSX' 12.0 and later"),
```

## Additional context

This is not fixed by https://github.com/dotnet/roslyn-analyzers/pull/7569 nor https://github.com/dotnet/roslyn-analyzers/pull/7644.

### LINKED ISSUE #7665: CA1416 with multiple guard attributes
### Analyzer

**Diagnostic ID**: warning CA1416: This call site is reachable on: 'macOS/OSX' 11.0 and later, 'tvos' 12.2 and later. 'TestType.PerformanceRating' is only supported on: 'tvos' 13.0 and later.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

```shell
$ .dotnet/dotnet --version
10.0.100-preview.3.25167.3
```

### Describe the bug

Incorrect CA1416 when using multiple guard attributes.

Test code (from provided unit test):

```cs
using System;
using System.Runtime.Versioning;

// Commenting out either the macOS or the tvOS attributes (in all three places) makes this work.
[assembly: SupportedOSPlatform ("macos12.0")]
[assembly: SupportedOSPlatform ("tvos12.2")]

partial class TestType {
    void DoSomething ()
    {
        if (IsAtLeastXcode11) {
            Console.WriteLine (PerformanceRating); // warning CA1416: This call site is reachable on: 'macOS/OSX' 11.0 and later, 'tvos' 12.2 and later. 'TestType.PerformanceRating' is only supported on: 'tvos' 13.0 and later.
        }
    }

    [SupportedOSPlatform ("macos11.0")]
    [SupportedOSPlatform ("tvos13.0")]
    public ulong? PerformanceRating { get; private set; }

    [SupportedOSPlatformGuard ("macos11.0")]
    [SupportedOSPlatformGuard ("tvos13.0")]
    internal static bool IsAtLeastXcode11 {
        get {
            return true;
        }
    }
}
```

One interesting thing here is that commenting out either all the tvOS attributes or all the macOS attributes makes the warning go away.

### Steps To Reproduce

Run this unit test: https://github.com/rolfbjarne/roslyn-analyzers/commit/8746314b744ba71bb87c2c30c49cda0895713491

### Expected behavior

Unit test should pass

### Actual behavior

Unit test fails with:

```
System.InvalidOperationException : Context: Diagnostics of test state
Mismatch between number of diagnostics returned, expected "0" actual "1"

Diagnostics:
// /0/Test0.cs(12,32): warning CA1416: This call site is reachable on: 'macOS/OSX' 11.0 and later, 'tvos' 12.2 and later. 'TestType.PerformanceRating' is only supported on: 'tvos' 13.0 and later.
VerifyCS.Diagnostic(PlatformCompatibilityAnalyzer.CA1416).WithSpan(12, 32, 12, 49).WithArguments("TestType.PerformanceRating", "'tvos' 13.0 and later", "'macOS/OSX' 11.0 and later, 'tvos' 12.2 and later"),
```

### PR DESCRIPTION
**Pull request by rolfbjarne**
## Summary

This PR fixes two independent `CA1416` (Platform compatibility) false positives in `PlatformCompatibilityAnalyzer`, adds systematic test coverage for the platform-attribute combination matrix that surfaced the second one, and works through five further `CA1416` issues reported against this analyzer.

| # | Bug | Fixed in |
|---|--------------------------------------------------------------------|----------------------------------|
| 1 | Multi-platform custom guards mishandled when the call site already covers one platform | `IsKnownValueGuarded` |
| 2 | An `ios` `Supported`/`Unsupported` pair naming the same version cancelled out on `ios` but stranded `maccatalyst` | `TryAddValidAttribute` |

Bug 1's fix also resolves a second reported issue, [roslyn-analyzers#7661](https://github.com/dotnet/roslyn-analyzers/issues/7661) — see [Other reported issues](#other-reported-issues) below.

---

## Bug 1 — multi-platform custom guards

`CA1416` reported false positives when a custom `[SupportedOSPlatformGuard]` (or `[UnsupportedOSPlatformGuard]`) declared **multiple platform attributes**, and the call site's own platform support already satisfied one or more of those guards.

### Repro

The simplest reproduction ([roslyn-analyzers#7665](https://github.com/dotnet/roslyn-analyzers/issues/7665)):

```csharp
[assembly: SupportedOSPlatform("macos12.0")]
[assembly: SupportedOSPlatform("tvos12.2")]

partial class TestType
{
    void DoSomething()
    {
        if (IsAtLeastXcode11)
        {
            Console.WriteLine(PerformanceRating); // ❌ spurious CA1416
        }
    }

    [SupportedOSPlatform("macos11.0")]
    [SupportedOSPlatform("tvos13.0")]
    public ulong? PerformanceRating { get; private set; }

    [SupportedOSPlatformGuard("macos11.0")]
    [SupportedOSPlatformGuard("tvos13.0")]
    internal static bool IsAtLeastXcode11 => true;
}
```

Inside the `if`, `PerformanceRating` is fully guarded, yet the analyzer warned. Tellingly, removing *either* the macOS or the tvOS attributes (from all three sites) made the warning disappear — the bug only surfaced with mixed-platform guards.

### Root cause

When processing a guard, the analyzer treated all of its platform attributes as a single unit. If the call site's existing support already covered one platform of a multi-platform guard, the analyzer mishandled the remaining platforms instead of evaluating each one on its own.

### Fix

In `IsKnownValueGuarded`, handle each platform of a mixed guard individually:

- **Supported guards already covered by the call site:** when a supported guard is suppressed because the call site's support already covers it, preserve the guarded flow (`RemoveOtherSupportsOnDifferentPlatforms`) instead of re-applying it as a call-site attribute. Gated by the new `IsPlatformSupportSuppressedByCallsite` helper.
- **Negated guards that can't be reached:** exclude a platform from a negated guard branch when that branch is unreachable at the call site's minimum version, while still analyzing platforms whose pre-version branch *is* reachable. Gated by the new `IsPlatformExcludedByCallsite` helper.

The `IsPlatformExcludedByCallsite` helper deliberately requires **every** call-site attribute to be an allow list. An earlier revision dropped that gate, which introduced a **false negative**: with a deny-list call site such as `[UnsupportedOSPlatform("macos12.0"), SupportedOSPlatform("macos15.0")]` the guarded branch is still reachable on every *other* platform, so excluding macOS would silently drop a macOS-only API's requirement. Both helpers carry comments explaining exactly which call-site shapes they are and are not sound for.

---

## Bug 2 — `maccatalyst` stranded by a cancelling `ios` attribute pair

Found while building out the combination matrix below.

`TryAddValidAttribute` mirrors every `ios` attribute onto `maccatalyst`, because `OperatingSystem.IsIOS()` carries `[SupportedOSPlatformGuard("maccatalyst")]` and Mac Catalyst is therefore treated as a superset of the iOS surface. That mirroring lived in the `else` branch of the check that drops a platform when `AddAttribute` returns `false` — so it was skipped in exactly the case where it mattered.

`AddAttribute` returns `false` when a `Supported` and an `Unsupported` attribute name the **same version**: they cancel out and the platform is dropped. So for:

```csharp
[UnsupportedOSPlatform("ios12.0")]
[SupportedOSPlatform("ios12.0")]
public static void Api() { }
```

the first attribute wrote `unsupported 12.0` to both `ios` **and** `maccatalyst`; the second cancelled `ios` and removed it, but never reached `maccatalyst`. The API was then reported as *"unsupported on: 'maccatalyst' 12.0 and later"* while being fully supported on iOS — and the warning was **unsilenceable**, since nothing `maccatalyst`-specific appears anywhere in the source.

Writing the identical pattern with `macos`/`maccatalyst` cancels cleanly, which is what makes this a bug rather than a deliberate quirk. That control ships as a test.

The fix cancels the mirrored attribute on the related platform too. It deliberately does **not** *create* a related entry in the cancel branch: with the primary platform cancelled there is nothing to mirror, and creating one could newly add `maccatalyst` support in a pathological ordering case.

---

## Test coverage

40 new tests. Beyond direct regression tests for the two fixes, this PR adds `PlatformCompatibilityAnalyzer.AttributeCombinationTests.cs`, a new file that works through the combination space of the platform attributes. Its header documents the seven dimensions:

| Dimension | Values |
|-------------------|--------------------------------------------------------------------------------|
| attribute kind | `SupportedOSPlatform` / `UnsupportedOSPlatform` / `ObsoletedOSPlatform` (+ both guard forms) |
| version | absent (treated as `0.0`), present, or *equal to another attribute's version* |
| multiplicity | one, two, or three attributes of the same kind on the same platform |
| platform relation | unrelated platforms versus the single related pair `ios -> maccatalyst` |
| declaration level | assembly, type, or member |
| call site shape | reachable everywhere, allow list, or deny list |
| spelling | `OSX` is an alias of `macos` |

Because Mac Catalyst is a superset of iOS, the `ios`/`maccatalyst` pair gets the most attention, and each inference test is paired with an explicit-attribute control so the inferred entry is shown to behave identically to a hand-written one.

Behaviours now locked in by tests, several of which had no coverage at all:

- Same-version cancel-out is **source-order dependent**: `(Unsupported, Supported)` removes the platform, `(Supported, Unsupported)` survives as a deny list.
- `Supported` and `Obsoleted` keep the **lowest** version, so an explicit `maccatalyst14.0` is swallowed by a `13.0` inferred from `ios`; `Unsupported` keeps only the **two lowest**, so a third, higher one is dropped.
- `Obsoleted`, the two-slot unsupported range, all three kinds at once, and assembly-level attributes all mirror correctly onto `maccatalyst`.
- Unrelated platforms never borrow each other's *version ranges* — the `macos`/`tvos` interaction that issue #7665 was originally reported for. (The header is explicit that this is **not** the same as claiming a macOS attribute has no effect on tvOS: a lone `SupportedOSPlatform` makes the API an allow list, so every platform without its own attribute becomes unsupported. That is intended and platform-agnostic, and is itself asserted by a test.)
- Custom *versioned* `ios` guards do not make a `maccatalyst`-only API reachable, matching the built-in `OperatingSystem.IsIOS()` behaviour.
- A member-level `Supported` cannot re-enable a platform that a type-level deny list turned off, and the inferred `maccatalyst` entry behaves exactly like an explicit one here.
- `OSX` folds into `macos`; a platform name starting with a digit is silently ignored in its entirety; four-part versions round-trip.

### Method

The combinations were not guessed. Scratch test files declared each combination as expecting *no* diagnostics, so the resulting failures printed the analyzer's **actual** output; the real tests were then written from that. Every fix was A/B validated by reverting the analyzer and re-running the new tests, which confirmed that exactly one test fails without the corresponding change — proving both that the fixes are narrowly scoped and that the regression tests are genuine.

---

## Other reported issues

The five other open `CA1416` issues filed against this analyzer were each reduced to a minimal repro and re-evaluated against this branch. The repros ship as `PlatformCompatibilityAnalyzer.ReportedIssuesTests.cs` so that the conclusions stay honest — including the ones that are *not* bugs, which would otherwise be silently re-litigated later.

| Issue | Verdict | Test |
|-------------------------------|----------------------------------------------|--------------------|
| [roslyn-analyzers#7661](https://github.com/dotnet/roslyn-analyzers/issues/7661) | **Fixed by this PR** | passing |
| [roslyn-analyzers#7392](https://github.com/dotnet/roslyn-analyzers/issues/7392) | Not a bug — feature request | passing |
| [roslyn-analyzers#7578](https://github.com/dotnet/roslyn-analyzers/issues/7578) | Still a bug | `[Ignore]`d |
| [#54066](https://github.com/dotnet/sdk/issues/54066) | Still a bug | `[Ignore]`d |
| [#51144](https://github.com/dotnet/sdk/issues/51144) | Still a bug | `[Ignore]`d |

**#7661 is fixed by this PR.** The combination matrix missed it because it pairs a *negated* guard with an `Unsupported` API, a shape none of the matrix cells covered. That it is genuinely fixed here — rather than having never reproduced — was established by A/B: the same test fails when the analyzer is reverted to `origin/main`.

**#7392 is not a bug.** The request is for the project's target framework version to act as a version *floor* for `CA1416`, so that `[SupportedOSPlatform("ios13.0")]` stops warning in a project targeting `net8.0-ios17.0`. The analyzer already honours such a floor when it is expressed as an assembly-level `SupportedOSPlatform` — a control test asserts exactly that — so this is a request for the SDK to emit that attribute from the TFM, not an analyzer defect.

The remaining three are genuine bugs, and each ships as a known-failure test marked `[TestMethod, Ignore("...")]` with a link to its issue. **They are deliberately not fixed here**: this PR is large enough already, and the three fixes touch merge and guard logic that deserves its own review. Landing the repros now means the fixes arrive as a diff that flips an `[Ignore]`, which is far easier to review than a fix and its test arriving together.

### Results

Full `Microsoft.CodeAnalysis.NetAnalyzers.UnitTests` assembly: **17 758 tests, 0 failures** (35 pre-existing skips, plus the 3 known-failure tests described above).

Fixes https://github.com/dotnet/roslyn-analyzers/issues/7665
Fixes https://github.com/dotnet/roslyn-analyzers/issues/7661

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.NetAnalyzers/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 1]
```diff
@@ -656,11 +670,34 @@ static bool IsKnownValueGuarded(
...
 
+            // A call site without a minimum supported version for the platform cannot prove the branch unreachable,
+            // so 'SupportedFirst' being null must not exclude the platform.
+            static bool IsPlatformExcludedByCallsite(
+                PlatformMethodValue value,
+                SmallDictionary<string, Versions>? callsiteAttributes)
+                => callsiteAttributes != null &&
+                    callsiteAttributes.TryGetValue(value.PlatformName, out Versions? attributes) &&
+                    attributes.SupportedFirst != null &&
+                    attributes.SupportedFirst.IsGreaterThanOrEqualTo(value.Version);
```
Comment (tannergooding) [inline src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.NetAnalyzers/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs]: For this and `IsPlatformSupportSuppressedByCallsite`, I always get confused by the rules.

In particular, `UnsupportedOSPlatform` documents itself as:
> If a version is specified, the API is only available in earlier versions.

While `SupportedOSPlatform` documents itself as:
>  If a version is specified, the API cannot be called from an earlier version

So what happens if a user specifies both `[UnsupportedOSPlatform("macos12")]` + `[SupportedOSPlatform("macos15")]`?

Do we have a clear test covering scenarios like this and should it be called out with a code comment/example explaining how such edges are expected to work?

**Comment by tannergooding** [comment]
The changes here generally look correct; but we do not have nearly enough documentation, examples, or tests covering the edge cases here and how they are expected to work. Especially when `Supported`, `Unsupported`, and `Obsoleted` get mixed or possibly used together in "invalid" combinations.

The docs cover some specific rules, but those aren't even fully self-consistent given real world usages; so I think this really deserves some explicit coverage and tests covering things. Both for singular platforms (only considering say windows) and for the more interesting platforms (like how macos, maccatalyst, and ios all oddly overlap and imply eachother in weird ways).

File: src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.NetAnalyzers/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.NetAnalyzers/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs]: > So what happens if a user specifies both `[UnsupportedOSPlatform("macos12")]` + `[SupportedOSPlatform("macos15")]`?

Fortunately we've never had an API re-introduced, so this is somewhat academic at the moment. That said, I believe it's documented [here](https://learn.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer#advanced-scenarios-for-attribute-combinations), where it says: "An API that was supported initially but unsupported (removed) in a later version is not expected to get re-supported in an even later version."

> Do we have a clear test covering scenarios like this and should it be called out with a code comment/example explaining how such edges are expected to work?

Good idea, I'll add some more tests.

**Comment by rolfbjarne** [comment]
@tannergooding Thanks for the review! I've added a significant number of new tests, which uncovered a few more bugs, which I've also fixed. I've also gone through existing issues in this area; one will be fixed with this PR, for the remaining ones I've also added unit tests that reproduces them (but disabled, because this PR's goal isn't to fix those issues - I'll do that in a separate PR later).
